src/lobby.rs

```diff
@@ -194,11 +194,26 @@ impl SharedLobbyState {
         state.active_contributor = ActiveContributor::None;
     }
 
+    #[allow(clippy::needless_collect)]
     pub async fn clear_lobby(&self, predicate: impl Fn(&SessionInfo) -> bool + Copy + Send) {
         let mut lobby_state = self.inner.lock().await;
-        lobby_state
+        let sessions_to_remove = lobby_state
             .sessions_in_lobby
-            .retain(|_, info| !predicate(info));
+            .iter()
+            .filter_map(|(id, info)| {
+                if predicate(info) {
+                    Some(id.clone())
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<_>>();
+        for id in sessions_to_remove {
+            let info = lobby_state.sessions_in_lobby.remove(&id);
+            if let Some(info) = info {
+                lobby_state.sessions_out_of_lobby.insert(id, info);
```

Comment:
As long as there are no timeouts/deadlines on those in the `out_of_lobby` list, then this should never cause a problem

Reply:
There is a separate timeout for these, but it's at 24hrs by default, which should not cause spurious failures.